lib/ansible/modules/network/nxos/nxos_file_copy.py

```diff
@@ -185,29 +231,121 @@ def transfer_file(module, dest):
+                child.sendline('yes')
+                child.expect('password')
+        elif index == 2:
+            child.sendline('yes')
+            child.expect('password')
+        child.sendline(module.params['remote_scp_server_password'])
+        fpt = module.params['file_pull_timeout']
+        # response could be that there is no space left on device,
+        # permission denied due to wrong user/password,
+        # remote file non-existent or success
+        index = child.expect(['No space', 'Permission denied', 'No such file', '#'], timeout=fpt)
+        if index == 0:
+            module.fail_json(msg='File copy failed due to no space left on the device')
+        elif index == 1:
+            module.fail_json(msg='Username/Password for remote scp server is wrong')
+        elif index == 2:
+            module.fail_json(msg='File copy failed due to remote file not present')
+    except pexpect.ExceptionPexpect as e:
+        module.fail_json(msg='%s' % to_native(e), exception=traceback.format_exc())
+
+    child.close()
+
+
 def main():
     argument_spec = dict(
-        local_file=dict(required=True),
-        remote_file=dict(required=False),
+        local_file=dict(type='str'),
+        remote_file=dict(type='str'),
         file_system=dict(required=False, default='bootflash:'),
         connect_ssh_port=dict(required=False, type='int', default=22),
+        file_pull=dict(type='bool', default=False),
+        file_pull_timeout=dict(type='int', default=300),
+        remote_scp_server=dict(type='str'),
+        remote_scp_server_user=dict(type='str'),
+        remote_scp_server_password=dict(no_log=True),
     )
 
     argument_spec.update(nxos_argument_spec)
 
-    module = AnsibleModule(argument_spec=argument_spec, supports_check_mode=True)
+    rt = [['remote_scp_server',
+           'remote_scp_server_user',
+           'remote_scp_server_password']]
+
+    module = AnsibleModule(argument_spec=argument_spec,
+                           required_together=rt,
+                           supports_check_mode=True)
+
+    file_pull = module.params['file_pull']
+    local_file = module.params['local_file']
+    remote_file = module.params['remote_file']
+    remote_scp_server = module.params['remote_scp_server']
 
-    if not HAS_PARAMIKO:
-        module.fail_json(
-            msg='library paramiko is required but does not appear to be '
-                'installed. It can be installed using `pip install paramiko`'
-        )
+    if file_pull:
+        if not HAS_PEXPECT:
+            module.fail_json(
+                msg='library pexpect is required but does not appear to be '
+                    'installed. It can be installed using `pip install pexpect`'
+            )
 
-    if not HAS_SCP:
-        module.fail_json(
-            msg='library scp is required but does not appear to be '
-                'installed. It can be installed using `pip install scp`'
-        )
+        if not remote_file:
```

Comment:
There is standard way to have these checks.
```
required_if = [("file_pull", True, ["remote_file", "remote_scp_server"]),
                      ("file_pull", False, ["local_file"])]

if file_pull: 
    check remote_file requirements
else:
    check_local_file requirements
```
Pass `required_if` to AnsibleModule(required_if=required_if) when creating the module object.